src/compiler/checker.ts

```diff
@@ -11083,10 +11096,25 @@ namespace ts {
             const parent = container && container.parent;
             if (parent && (isClassLike(parent) || parent.kind === SyntaxKind.InterfaceDeclaration)) {
                 if (!hasModifier(container, ModifierFlags.Static) &&
-                    (container.kind !== SyntaxKind.Constructor || isNodeDescendantOf(node, (<ConstructorDeclaration>container).body!))) {
+                    (!isConstructorDeclaration(container) || isNodeDescendantOf(node, container.body))) {
                     return getDeclaredTypeOfClassOrInterface(getSymbolOfNode(parent as ClassLikeDeclaration | InterfaceDeclaration)).thisType!;
                 }
             }
+
+            // inside x.prototype = { ... }

```

Comment:
getThisType and checkThisExpression should use the same code to look up the class' `thisType`, but do not. In particular, checkThisExpression's code, while more complete, is less efficient than this code since it uses `checkExpressionCached` instead of just grabbing symbol.parent. Since it's a separate change, I'll fix it in a followup PR.